FuelLabs/fuel-indexer
# enhancement: support graphql-playground

Closes #646 

## Changelog
• Adds the crates `async-graphql` and `async-graphql-axum`.
• Adds the new api route `/playground/:namespace/:identifier`.
• Adds new handler function for the above route `gql_playground`.
• Appends data field to query responses so that playground renders response correctly. 
• Updates tests to assert against the new response. 
• Updates docs with new query.
• Updates the response type for handler function `query_graph` to `GraphQLRequest`.

## Context
When initially testing querying in Postman, I found that our api would only accept a request of raw string. When I tried to use Postmans GraphQL testing request type, our server could not parse the request. I was able to reproduce this with the Playground, the playground was not functioning as it formats its requests to Graphql Request types. It was safe to assume that any GraphQL client would also not work. 

I looked at some other repositories using graphql with the playground they implemented the `async_graphql_axum::GraphQLRequest` type. After doing this, our server was able to parse the queries from a graphQL client. 

If we want prettier responses with the playground, we would have to update our response types to `async_graphql_axum::GraphQLResponse`, but I thought this was unnecessary for now. 

## Testing
1. Run your local fuel-node `cargo run --bin fuel-node`
1. Spin up the test indexer  `cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml`
2. In your browser go to `http://localhost:29987/api/playground/fuel_indexer_test/index1` (you will see some error on initial render, ignore it).
3. In the GraphQL Playground run the query: 
```
query { 
    block 
    { 
        id 
        height 
        timestamp 
        }
}
```
You should receive the response. (If this fails you may need to trigger an event via the web-api so that the DB has some data to query).
```
{
    "data": [
        {
            "height": 1,
            "id": 3919316288814604599,
            "timestamp": 1680799333
        },
        {
            "height": 2,
            "id": 3832952940330561849,
            "timestamp": 1680799423
        },
        {
            "height": 3,
            "id": 3977631069887424055,
            "timestamp": 1680799493
        },
        {
            "height": 1,
            "id": 4121747375410198369,
            "timestamp": 1680800681
        },
        {
            "height": 2,
            "id": 7378647927350505569,
            "timestamp": 1680800974
        },
        {
            "height": 1,
            "id": 4135211972506432054,
            "timestamp": 1680801051
        },
        {
            "height": 2,
            "id": 3702856512725070640,
            "timestamp": 1680801341
        },
        {
            "height": 1,
            "id": 3474306528200385894,
            "timestamp": 1680811403
        }
    ]
}
```
For more optional testing you can try deploying a second indexer and using that namespace and identifier to query through the playground, if you like.
